### ISSUE TITLE
scvi-tools vs scvi for a pre-trained totalvi model

### ISSUE THREAD

**Issue by stkachev** [YosefLab/scvi-tools#861]
Dear authors!

First of all let me thank you for this great tool that makes our CITE-Seq data integration a lot easier!

From the past I have a pre-trained model for which I used scvi 0.6.7. The actual data are CD3/CD28 treated PBMC so you would expect a lot of activated T cells but also with some B cells and NK cells. And so if I open that model using old API, get latent data and feed it to scanpy to do the clustering I get roughly what I expect:

![umaps_scvi](https://user-images.githubusercontent.com/16946837/99459722-8e590380-28fc-11eb-86f0-77b62a4ec8f8.png)

Now if I try to do the same using new API from scvi-tools there is a somewhat less structure than before:

![umaps_scvi_tools](https://user-images.githubusercontent.com/16946837/99459806-b183b300-28fc-11eb-8aff-14f889e2e035.png)

One aspect in particular are cells with high total_adt_counts and not-so-high total_umi that in previous API form a separate cluster, both visually and by clustering algorithm in scanpy. We think these are actually dying cells that trap antibodies through developing pores in their membranes. And so we would expect them to form a separate cluster. We do see it scvi 0.6.7 but not in scvi-tools 0.7.1. Do you have any idea what may be causing this?

Best,
Sasha,

**Comment by adamgayoso** [comment]
Hi @stkachev, it would be helpful if you provided scripts with how you ran totalVI in each scenario.

**Comment by stkachev** [comment]
These are not scripts really but more like pieces of python code that I was cutting-and-pasting into python console.

Anyway, starting from reading a pre-trained model.

Old way:

```
trainer.model.load_state_dict(torch.load('model.pkl',map_location=torch.device('cpu')))
trainer.model.eval()

full_posterior = trainer.create_posterior(type_class=TotalPosterior)
full_posterior = full_posterior.update({"batch_size":32})

latent_mean, batch_index, label, library_gene = full_posterior.get_latent()

post_adata = anndata.AnnData(X=dataset.X)
...
sc.pp.neighbors(post_adata, use_rep="X_totalVI", n_neighbors=30, metric="correlation")
sc.tl.umap(post_adata, min_dist=0.5)
sc.tl.leiden(post_adata, key_added="leiden_totalVI", resolution=0.8)
...
```

New way:

```
totalvae.model.load_state_dict(torch.load('model.pkl',map_location=torch.device('cpu')))
totalvae.model.eval()
totalvae.is_trained_ = True

dataset.obsm["X_totalVI"] = totalvae.get_latent_representation(batch_size=32)

sc.pp.neighbors(dataset, use_rep="X_totalVI", n_neighbors=30, metric="correlation")
sc.tl.umap(dataset, min_dist=0.5)
sc.tl.leiden(dataset, key_added="leiden_totalVI", resolution=0.8)
...
```

**Comment by adamgayoso** [comment]
Could you also provide the code used to train each of the models? For example, were any non-default parameters set?

**Comment by adamgayoso** [comment]
Finally, is model.pkl the model trained with the scvi package?

I can see how there would be an issue if you're using scvi-tools version 0.8.0b0, where we have different defaults.

**Comment by stkachev** [comment]
There is only one model which I trained with scvi 0.6.7. I am just trying to read it with new API. 

And the code I used at a time is like this:

```
totalvae = TOTALVI(dataset.nb_genes,
                   len(dataset.protein_names),
                   n_batch=dataset.n_batches,
                   dropout_rate_encoder=0.1,
                   dropout_rate_decoder=0.1,
                   n_latent=40
)

use_cuda = True
lr = 4e-3
n_epochs = 5000

live_alpha_prior = torch.tensor([2.2552, 2.4116, 2.6447, 2.9225, 1.2948, 1.6462, 2.6863, 2.4364,
                                 0.8724, 0.7859, 0.8791, 0.8892, 0.7906, 0.7674, 0.9913, 1.5589]) * 1

live_log_beta_prior = torch.tensor([ -0.5276, -0.1397, -0.1595, -0.3373, -0.7328, -0.7030, -0.7403, -0.6343,
                                     -0.5226, -1.2027, -0.7263, -1.1481, -1.3778, -1.5900, -0.8392, 0.2667])

dsp_alpha_prior = torch.tensor([1.5521, 2.5691, 3.5860, 1.5081, 1.1143, 0.9958, 1.4366, 1.4056,
                                2.3353, 1.8306,1.1729, 2.3101, 2.1444, 3.2974, 2.7881, 1.0306]) * 1

dsp_log_beta_prior = torch.tensor([-0.44259, -0.15854, -0.09064, -0.55507, -0.67932, -0.84542, -0.52627, -0.60193,
                                   -0.85668, -0.70607, -0.91427, -0.57187, -0.72504, -0.82441, -0.68233, -1.04495])

dsp2x_alpha_prior = torch.tensor([1.561, 2.572, 3.550, 1.532, 1.135, 1.014, 1.435, 1.469,
                                  1.905, 1.390, 1.084, 2.009, 1.970, 3.236, 2.770, 0.938]) * 1

dsp2x_log_beta_prior = torch.tensor([-0.4287, -0.1132, -0.0181, -0.5358, -0.6382, -0.8167, -0.4954, -0.6021,
                                     -0.5871, -0.6818, -0.9295, -0.3672, -0.4991, -0.5762, -0.4106, -1.1230])

lyuowort_dsp_alpha_prior = torch.tensor([2.201, 2.866, 3.813, 2.207, 1.390, 1.347, 2.215, 2.195,
                                         2.036, 1.671, 1.298, 2.203, 1.842, 2.408, 2.861, 1.072]) * 1

lyuowort_dsp_log_beta_prior = torch.tensor([-0.27732,-0.05379,0.02637, -0.43335,-0.50349, -0.58787, -0.34346, -0.50034,
                                            -0.77582,-0.69673,-0.83200,-0.51550,-0.66313, -0.75340, -0.61522, -0.96875])

lyuowort_dsp2x_alpha_prior = torch.tensor([2.090, 2.850, 3.701, 2.082, 1.349, 1.268, 2.097, 2.089,
                                           1.819, 1.442, 1.361, 1.960, 1.705, 2.229, 2.766, 1.046]) * 1

lyuowort_dsp2x_log_beta_prior = torch.tensor([-0.30783,-0.05974,-0.01075,-0.45977,-0.51401,-0.62937,-0.35806, -0.52412,
                                              -0.56125,-0.61938,-0.65507,-0.35087,-0.50468,-0.56099,-0.45713, -0.92825])

totalvae.background_pro_alpha[:,0].data.copy_(live_alpha_prior)
totalvae.background_pro_log_beta[:,0].data.copy_(live_log_beta_prior)
totalvae.background_pro_alpha[:,1].data.copy_(dsp_alpha_prior)
totalvae.background_pro_log_beta[:,1].data.copy_(dsp_log_beta_prior)
totalvae.background_pro_alpha[:,2].data.copy_(dsp2x_alpha_prior)
totalvae.background_pro_log_beta[:,2].data.copy_(dsp2x_log_beta_prior)
totalvae.background_pro_alpha[:,3].data.copy_(lyuowort_dsp_alpha_prior)
totalvae.background_pro_log_beta[:,3].data.copy_(lyuowort_dsp_log_beta_prior)
totalvae.background_pro_alpha[:,4].data.copy_(lyuowort_dsp2x_alpha_prior)
totalvae.background_pro_log_beta[:,4].data.copy_(lyuowort_dsp2x_log_beta_prior)

trainer = TotalTrainer(
    totalvae,
    dataset,
    train_size=0.90,
    test_size=0.10,
    use_cuda=use_cuda,
    frequency=1,
    batch_size=512,
    use_adversarial_loss=True,
)

trainer.train(lr=lr, n_epochs=n_epochs)
torch.save(trainer.model.state_dict(),'model.pkl')
torch.save(trainer.history,"trainer_history.pkl")

```

**Comment by adamgayoso** [comment]
Everything looks normal so far. Could you share how you initialize `totalvae` in scvi-tools?

**Comment by stkachev** [comment]
Sure, here it is

```
use_cuda = True

totalvae = scvi.model.TOTALVI(dataset,
                              dropout_rate_encoder=0.1,
                              dropout_rate_decoder=0.1,
                              n_latent=40,
                              use_cuda=use_cuda
)

```

**Comment by stkachev** [comment]
And then I proceed to reading it with `load_state_dict`...

**Comment by adamgayoso** [comment]
So maybe one of two things is happening:

1) 0.8.0b0 is accidentally being used, where the model would be initialized with `latent_distribution = "normal"`
2) The batches in your data are being mapped to different values from those used to train your model.

you can inspect `adata.obs["_scvi_batch"]` to see that things are what you expect.

**Comment by adamgayoso** [comment]
by the way, I believe you had raised the issue about the initialization of the alpha and beta. I have a new approach in 0.8.0b0 that you can read a bit about in the release notes 

https://www.scvi-tools.org/en/latest/release_notes/v0.8.0.html

Note however that you won't be able to load models like you are doing here in 0.8.0 because some of the key names changed.

**Comment by stkachev** [comment]
totalvae says latent distribution is `ln`. 

For the batches indeed there is maybe a mismatch. 

```
>>> dataset.obs[['Sample','_scvi_batch']]
                              Sample  _scvi_batch
AAACCCAAGGACACTG-1-0            live            4
AAACCCAGTACCCACG-1-0            live            4
AAACCCAGTAGTCGTT-1-0            live            4
AAACCCAGTATTGCCA-1-0            live            4
AAACCCAGTCCATAGT-1-0            live            4
...                              ...          ...
TTTGTTGAGATGCCGA-1-4  dsp2x_LyUoWort            2
TTTGTTGAGCATAGGC-1-4  dsp2x_LyUoWort            2
TTTGTTGCAGTTGTCA-1-4  dsp2x_LyUoWort            2
TTTGTTGTCTAAGGAA-1-4  dsp2x_LyUoWort            2
TTTGTTGTCTAGTCAG-1-4  dsp2x_LyUoWort            2

[28929 rows x 2 columns]

```

I wonder what it was using in  scvi 0.6.7 version. If I set `dataset.obs[['_scvi_batch']]` to what I think it was (such as "0" for live) umap plots do not improve unfortunately.

I will try to re-train everything using scvi-tools with the same parameters that I used for scvi 0.6.7, maybe I will get a comparable result to what I used to have.

And yeah it's great it will be possible to do empirical alpha and beta initialization!

**Comment by adamgayoso** [comment]
> I will try to re-train everything using scvi-tools with the same parameters that I used for scvi 0.6.7, maybe I will get a comparable result to what I used to have.

It would be fantastic if you could confirm this. Everything should indeed be the same. We also made some speed improvements related to early stopping.

> If I set dataset.obs[['_scvi_batch']] to what I think it was (such as "0" for live) umap plots do not improve unfortunately.

I would also make sure all other features (genes and proteins) are in the same order in the two respective dataset objects.

**Comment by stkachev** [comment]
It's a little better if I retrain it. However with the 0.6.7 version I feel like there are still a little more details. For example previously next to B cells cluster (island at the bottom) there are two smaller sets of cells, one to the left and another to the right. One of those are B cells that we only see in one of the batches and another one are small number of myeloid cells that survived our CD3/CD28 treatment.

![umaps](https://user-images.githubusercontent.com/16946837/99545384-534de300-2983-11eb-8371-f8aff5c39e18.png)

Also I tried to re-run the training but I got identical result between first and second run. Previously I could run it multiple times and then choose which run looks best.

**Comment by adamgayoso** [comment]
You can run `scvi.settings.seed = SEED` before each run to change the initial seed. Is this 0.7.1 or 0.8.0b0?

In 0.7.1, there were no changes to the actual model from the old scvi package, so any difference is either due to hyperparam differences, or run to run stochasticity.

**Comment by stkachev** [comment]
It is 0.7.1 and yes I will try a different seed. All hyperparameters are the same already.

Does this seed also affects `get_latent_representation()`? In 0.6.7 re-running `get_latent()` is giving slightly different results, I wonder what is the new behavior.

**Comment by adamgayoso** [comment]
on 0.7.1, `get_latent_representation()` should give different results each time it is called, even with the seed set. Does it not do that for you?

**Comment by stkachev** [comment]
It turned out I had a bug in my protein background initialization: I was assigning my pre-computed values to random batches in 0.7.1 version. With that fixed and also using `scvi.settings.seed` to try out different runs (I tried five) gave me results that are not too different from what I got in 0.6.7:

![umaps](https://user-images.githubusercontent.com/16946837/99610737-5e832c00-29e0-11eb-9887-92e63cadc378.png)

And re-running `get_latent_representation()` indeed gave me some variability in results.

**Comment by adamgayoso** [comment]
Fantastic! If you're curious, you should be able to pip install our new beta to test out the new empirical initialization.

**Comment by stkachev** [comment]
I tried out the new version and compared your new empirical initialization with my own hard coded values. And I tried three different seeds for each. For each seed difference between these initializations is kind of subtle. Some seeds are definitely better than others though.

Also NK cells are now better visible in the UMAP plots which is good. 

And of course trainer history curve is a lot less choppy then it was with training error dropped significantly.

**Comment by adamgayoso** [comment]
The convergence might also be due to us using observed library size, instead of treating it as a latent variable.

**Comment by stkachev** [comment]
Interesting! In the 0.6.7 version I was actually plotting observed library sizes versus scvi derived ones and was getting something very close to a straight line (color coding in the plot below is by sample).

Is it a fixed behavior now that library sizes are non-trainable or there is some kind of option to have them trainable?

![library_sizes](https://user-images.githubusercontent.com/16946837/100773355-c0736680-33ce-11eb-9ec3-2125877b89b3.png)

**Comment by adamgayoso** [comment]
You can pass `use_observed_lib_size=False` to the `TOTALVI` init method to revert to the old functionality.
